### PR TITLE
Support per buffer pool watermark polling mode

### PR DESCRIPTION
**Pull request by wendani**
**What I did**

Buffer pool watermark design is built on a READ_AND_CLEAR polling mode at the syncd level. We observe some SAI implementation is missing the clear_stats operation support for buffer pool watermark, either simply not coded yet or having hardware limitation. In such cases, we have mismatch between the actual polling behavior and what a user perceives from FLEX_COUNTER_DB read from FLEX_COUNTER_GROUP_TABLE.

To have a consistent view, we propose per buffer pool watermark stats polling mode at the orchagent level when not all buffer pools support clear_stats operation on a switch. The situation is detected by issuing clear_stats operation at the orchagent level to probe the capability over all pools first. If this is truly the case, we then do not set "STATS_MODE" field in "FLEX_COUNTER_GROUP_TABLE:BUFFER_POOL_WATERMARK_STAT_COUNTER", but set it to the per buffer pool table "FLEX_COUNTER_TABLE:BUFFER_POOL_WATERMARK_STAT_COUNTER:oid:<buffer_pool_oid>".

**Why I did it**

**How I verified it**
Tested on dut

```
127.0.0.1:6379[5]> hgetall "FLEX_COUNTER_TABLE:BUFFER_POOL_WATERMARK_STAT_COUNTER:oid:0x180000000005f3"
1) "BUFFER_POOL_COUNTER_ID_LIST"
2) "SAI_BUFFER_POOL_STAT_WATERMARK_BYTES"
3) "STATS_MODE"
4) "STATS_MODE_READ_AND_CLEAR"
127.0.0.1:6379[5]> hgetall "FLEX_COUNTER_TABLE:BUFFER_POOL_WATERMARK_STAT_COUNTER:oid:0x180000000005f4"
1) "BUFFER_POOL_COUNTER_ID_LIST"
2) "SAI_BUFFER_POOL_STAT_WATERMARK_BYTES"
3) "STATS_MODE"
4) "STATS_MODE_READ_AND_CLEAR"
127.0.0.1:6379[5]> hgetall "FLEX_COUNTER_TABLE:BUFFER_POOL_WATERMARK_STAT_COUNTER:oid:0x180000000005f5"
1) "BUFFER_POOL_COUNTER_ID_LIST"
2) "SAI_BUFFER_POOL_STAT_WATERMARK_BYTES"
3) "STATS_MODE"
4) "STATS_MODE_READ"
```

```
"FLEX_COUNTER_GROUP_TABLE:BUFFER_POOL_WATERMARK_STAT_COUNTER"
1) "BUFFER_POOL_PLUGIN_LIST"
2) "ca4aef2fce419033b7cd2ca9c787753198880af3"
3) "POLL_INTERVAL"
4) "10000"
5) "FLEX_COUNTER_STATUS"
6) "enable"
```

```
$ redis-cli -n 2 hgetall "COUNTERS_BUFFER_POOL_NAME_MAP"
1) "egress_lossless_pool"
2) "oid:0x180000000005f3"
3) "egress_lossy_pool"
4) "oid:0x180000000005f4"
5) "ingress_lossless_pool"
6) "oid:0x180000000005f5"
```

```
$ watermarkstat -t buffer_pool
Shared pool maximum occupancy:
                 Pool    Bytes
---------------------  -------
 egress_lossless_pool  1073904
    egress_lossy_pool        0
ingress_lossless_pool  1018160
```

**Details if related**

Depend on: https://github.com/Azure/sonic-sairedis/pull/485

### CODE REVIEW

File: orchagent/bufferorch.cpp [hunk 1]
```diff
@@ -170,14 +165,58 @@ void BufferOrch::generateBufferPoolWatermarkCounterIdList(void)
...
+    // Invoke the SAI clear_stats API per pool to query the capability from the API call return status
+    unsigned int noWmClrCapability = 0;
+    unsigned int bitMask = 1;
+    for (const auto &it : *(m_buffer_type_maps[CFG_BUFFER_POOL_TABLE_NAME]))
+    {
+        sai_status_t status = sai_buffer_api->clear_buffer_pool_stats(
+                it.second,
+                static_cast<uint32_t>(bufferPoolWatermarkStatIds.size()),
+                reinterpret_cast<const sai_stat_id_t *>(bufferPoolWatermarkStatIds.data()));
+        if (status !=  SAI_STATUS_SUCCESS)
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: >status !=  SAI_STATUS_SUCCESS [](start = 12, length = 29)

check for SAI_STATUS_NOT_IMPLEMENTED?

File: orchagent/bufferorch.cpp [hunk 2]
```diff
@@ -170,14 +165,58 @@ void BufferOrch::generateBufferPoolWatermarkCounterIdList(void)
         statList.pop_back();
     }
 
-    vector<FieldValueTuple> fvTuples;
-    fvTuples.emplace_back(BUFFER_POOL_COUNTER_ID_LIST, statList);
+    // Some platforms do not support buffer pool watermark clear operation on a particular pool
+    // Invoke the SAI clear_stats API per pool to query the capability from the API call return status
+    unsigned int noWmClrCapability = 0;
+    unsigned int bitMask = 1;
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: Why use a bit mask? What is the length of m_buffer_type_maps[CFG_BUFFER_POOL_TABLE_NAME]?

Comment (wendani) [reply orchagent/bufferorch.cpp]: It depends on the configuration. In our use case, it will be 3.

**Comment by kcudnik** [comment]
test this please

**Comment by lguohan** [comment]
retest this please

**Pull request merged**